### DISCUSSION TITLE
route implicit binding is fucked up

### DISCUSSION THREAD

**Discussion by Manu-sh** [livewire/livewire#3110] [7. Bugs (archived)]
I had already open this issue with another label but it disappeared so i will be more short and very less detailed... this doesn't work (it raise an error because $customer is null). If you need more details, ask.

```php
    public function mount(?Customer $customer) {
        $this->customer = $customer ?? new Customer();
    }
```

And should work because this is perfectly fine in laravel 5,6,7,8

```php
    Route::get('/customers/{customer?}', function (?Customer $customer = null) {
        $customer = $customer ?? new Customer();
        
        $customer->fill( ...);
        $customer->saveOrFail();
    });
```

```log
lluminate\View\ViewException:
Typed property App\Http\Livewire\Customer\Edit::$customer must be an instance of App\Models\Customer, null used (View: /home/user/fatturazione/resources/views/pages/customer/card.blade.php)

  at /home/user/fatturazione/vendor/livewire/livewire/src/LifecycleManager.php:90
  at Illuminate\View\Engines\CompilerEngine->handleViewException()
     (/home/user/fatturazione/vendor/livewire/livewire/src/ComponentConcerns/RendersLivewireComponents.php:73)
      ....
```

**Comment by austenc** [comment]
Please use the provided issue template. What version of Livewire are you using?

I haven't tested, but will it work if you do it like this? 
```
public function mount(?Customer $customer = null) {
```

I notice you have a default of null in your route definition, but not in the mount method, so you might try that. Assuming you are using full-page components here, what does the livewire route look like?

**Comment by hxnk** [comment]
I think this relates to issue #2402. I have the same issue.

Issues were introduced in release [v2.3.9](https://github.com/livewire/livewire/releases/tag/v2.3.9), so I'm currently fixed on v2.3.8.

**Comment by Manu-sh** [comment]
@austenc 

```
    "name": "livewire/livewire",
            "version": "v2.3.14",
```

**Comment by austenc** [comment]
@hxnk The issue you referenced seems to be about binding on non-mount methods, whereas this one is related to `mount`. 

@Manu-sh did you try my other suggestion? 
Is this a full page component? If so, what does the route for it look like?

**Comment by Manu-sh** [comment]
@austenc 

which is supposed to obtain is clear assuming i have
`customer id=1 in my db`

`card/customer/1 => edit mode`
`card/customer => creation mode`

i dont care really much about `card/customer/100` results,

**i have deleted this component however was something like**
```
// c/card.blade.php 
{{-- html --}}
<...  customer:="$customer" />
{{-- html --}}
```

where customer was injected from a regular laravel controller like every normal view
so:
```
Controller {
   pub fn card(?Customer $customer = null) {
      return view('c.card', compact('customer));
   }
}
```


and... if you remove `?Customer` from mount typehint everything works fine
so this work:

```
public function mount($customer) {
   $this->customer  = $customer ?? null;
}
```

**Comment by austenc** [comment]
@Manu-sh I believe all that you need to add is `= null` to your mount method as you had in the normal Laravel route in your first example.

**Comment by Manu-sh** [comment]
> @Manu-sh I believe all that you need to add is `= null` to your mount method as you had in the normal Laravel route in your first example.

is the same thing, it always crash i have tried

**Comment by austenc** [comment]
@Manu-sh please provide an example repository that depicts this problem and I will be happy to investigate.

**Comment by Manu-sh** [comment]
the problem is simple, the livewire typehint doesn't tollerate empty results set @austenc

**Comment by mokhosh** [comment]
This is working perfectly fine for me:
```php
// component
public User $user;

public function mount(User $user)
{
    $this->user = $user;
}

// routes
Route::get('user/{user?}', UserEditor::class);
```

**Comment by Manu-sh** [comment]
@mokhosh what if you forwarding this from a parent view?

```
<body>
  <h1>{{ $user->name }}</h1>
  <... $user="$user" />
</body>
```

**Comment by austenc** [comment]
Thanks @mokhosh, in my testing of this issue it worked correctly too. However, it looks like @Manu-sh isn't using a full-page component (and route). 

I'll be happy to investigate with an example repository. Otherwise, maintainers are left guessing about where the problem lies. @Manu-sh is that something you can provide?

**Comment by mokhosh** [comment]
OK, first of all you can achieve what you want like this:

```php
// component
public ?User $user;

public function mount()
{
    $this->user ??= new User;
}

// component view
<div>{{ $user->name }}</div>

// parent view
<livewire:user-editor :user="auth()->user()"/>
```

If you put `?User $user` parameter in mount method you will get an error though, which I think is a bit confusing.

Also, as a friend, the aggressive title and the way the question is posed is not really helping.

**Comment by Manu-sh** [comment]
@mokhosh i altready have found a workaround (just remove every typehint that's seems work except for the fact that i still gettings errors like if required fields where not filled) and however in this moment i still using plain laravel because i have no time, im pretty sure that this is an issue, and i think that the problem is that **some code that use reflection has logic errors**, I'm not the only user to report this, the title it's just from some user (me) that it's annoyed nothing more nothing less, don't take it personal I don't even know you.

**Comment by LooxisDev** [comment]
Is there anything new to this issue? I have still the same problem like described here:

https://github.com/livewire/livewire/issues/2696

I can't use any wire:click method, because it retrieves the model with the default id column, instead of my implicit route model binding
